WIP: Implement CRI stats collection

Mo-Fatah opened the pull request:
#### What type of PR is this?
/kind api-change
/kind feature

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

<!--
Fixes #
or
None
-->

#### Special notes for your reviewer:
This is a continuation of Sohan's effort in https://github.com/cri-o/cri-o/pull/7256 based on the recent cgroup manager changes in https://github.com/cri-o/cri-o/pull/7658

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Implement CRI container and sandbox stats collection 
```

haircommander commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,445 @@
+
+func findExistingContainerMetric(containerMetrics []*types.ContainerMetrics, containerID string) *types.ContainerMetrics {
+	for _, cm := range containerMetrics {
+		if cm.ContainerId == containerID {
```
hm I'm not sure about looping through the container metrics to save on cpu time of recreating the structure. It may be better to keep it simple to start and iterate later if needed

haircommander commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,445 @@
+
+// store metricdescriptors statically at startup, populate the list
+func (ss *StatsServer) PopulateMetricDescriptors(includedKeys []string) map[string][]*types.MetricDescriptor {
+	// TODO: add default container labels
```
is this TODO still legit?

haircommander commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,445 @@
+}
+
+func sandboxBaseLabelValues(sb *sandbox.Sandbox) []string {
+	// TODO FIXME: image?
```
(I wrote this TODO) What does cadvisor report as the image for the POD container?

Mo-Fatah replied:
I tried to look it up but couldn't find what they are using exactly as image name for POD containers. They seem to use the image name from the container info ([here](https://github.com/google/cadvisor/blob/570d8a7eea0a7669d0a1d024190010a200bbef7b/metrics/prometheus.go#L1782)  and [here](https://github.com/google/cadvisor/blob/570d8a7eea0a7669d0a1d024190010a200bbef7b/metrics/prometheus.go#L1807) but not sure how they differentiate between a container and a pod, and what image name exactly they are using for pods. 

my suggestion would be using the sandbox name defined in the sandbox metadata (e.g. [here](https://github.com/cri-o/cri-o/blob/345b566b48ddd45d400f7a14fff5bde7caa6ab30/test/testdata/sandbox_config.json#L3))

haircommander commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,445 @@
+			if existingMetric != nil {
+				existingMetric.Value = &types.UInt64Value{Value: v.value}
+			} else {
+				if metric == NetworkMetrics {
```
following up on the comment above: I think for now we should keep it simples and just continually replace the slices. We could get fancier in the future if we deem it necessary, but I think we're prone to issues doing so

haircommander commented:
great work here @Mo-Fatah ! I am leaning towards dropping the optimzations where we replace a slice in place. I think there will be an advantage to doing so in the future, but let's keep it simple to start. I also would love to see some unit and integration tests here to verify contents. If you don't have the bandwidth to add those let us know!

kannon92 commented on test/cri-metrics.bats:
```diff
@@ -0,0 +1,145 @@
+[crio.stats]
+collection_period = 0
+included_pod_metrics = [
+    "network",
```
@haircommander should we have processes here also?

haircommander replied:
yeah I think so

Mo-Fatah replied:
it should be yeah. but they are not yet supported in the collection implementation.

haircommander commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
+	var metrics []*types.Metric
+	for _, m := range ss.Config().IncludedPodMetrics {
+		switch m {
+		case "cpu":
```
note to self: verify how the kubelet correlates the metrics descriptors  to the metrics we return. It's possible we may need to be very dilligent about the ordering of which metrics we're including

haircommander replied:
actually, if we use the same list here and [here](https://github.com/cri-o/cri-o/pull/7796/files#diff-dbba0afd4b7f28a68d1393c2e4bd53d9e92a38b691b6e3bba1b8b87d9a709699R10) then we should be okay, as long as the actual metrics underneath are also in the same order

haircommander replied:
wait but if we put them in a [map](https://github.com/cri-o/cri-o/pull/7796/files#diff-dbba0afd4b7f28a68d1393c2e4bd53d9e92a38b691b6e3bba1b8b87d9a709699R11) and then transfer that to a slice, we'll lose the ordering...

haircommander commented on internal/oci/runtime_vm.go:
```diff
@@ -1014,6 +1014,10 @@ func metricsV1ToCgroupStats(ctx context.Context, m *cgroupsV1.Metrics) *cgmgr.Cg
 			SwapUsage:       m.Memory.Swap.Usage,
 			SwapLimit:       m.Memory.Swap.Limit,
+			// There is also m.Memory.TotalMappedFile
+			// TODO: See which one is the best to use here
```
who wrote this TODO, me or you?

Mo-Fatah replied:
I wrote it. Containerd metrics has two values: `TotalMappedFile` and `MappedFile`, and I wasn't sure which one is best to put here. I put the TODO as a note for anyone who might think there a wrong value reported

haircommander commented on server/metric_descriptors_list.go:
```diff
@@ -2,12 +2,23 @@ package server
 // ListMetricDescriptors lists all metric descriptors
 func (s *Server) ListMetricDescriptors(ctx context.Context, req *types.ListMetricDescriptorsRequest) (*types.ListMetricDescriptorsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "method unimplemented")
+	includedKeys := []string{"cpu", "memory", "network"} // You can modify this as needed
```
this should be the same list as https://github.com/cri-o/cri-o/pull/7796/files#diff-b9141e04c913bc0a793240448278d523cd5b1bd22cf036b9a0f3bcf1e32113c3R214

kwilczynski commented on docs/crio.8.md:
```diff
@@ -401,7 +407,7 @@ crio [GLOBAL OPTIONS] command [COMMAND OPTIONS] [ARGUMENTS...]
 **--signature-policy-dir**="": Path to the root directory for namespaced signature policies. Must be an absolute path. (default: "/etc/crio/policies")
 
-**--stats-collection-period**="": The number of seconds between collecting pod and container stats. If set to 0, the stats are collected on-demand instead. (default: 0)
+**--stats-collection-period**="": The number of seconds between collecting pod and container stats. If set to 0, the stats are collected on-demand instead. This option is deprecated and will be removed in the future. (default: 0)
```
```suggestion
**--stats-collection-period**="": The number of seconds between collecting pod and container stats. If set to 0, the stats are collected on-demand instead. DEPRECATED: This option will be removed in the future. (default: 0)
```

kwilczynski commented on docs/crio.conf.5.md:
```diff
@@ -528,7 +528,13 @@ The `crio.metrics` table containers settings pertaining to the Prometheus based
 
 **stats_collection_period**=0
-  The number of seconds between collecting pod and container stats. If set to 0, the stats are collected on-demand instead.
+  The number of seconds between collecting pod and container stats. If set to 0, the stats are collected on-demand instead.**Deprecated:** this option will be removed in future versions in favor of `collection_period`.
```
```suggestion
  The number of seconds between collecting pod and container stats. If set to 0, the stats are collected on-demand instead. **DEPRECATED:** This option will be removed in the future. Please use `collection_period` instead.
```

kwilczynski commented on docs/crio.8.md:
```diff
@@ -286,6 +290,8 @@ crio [GLOBAL OPTIONS] command [COMMAND OPTIONS] [ARGUMENTS...]
 
 **--imagestore**="": Store newly pulled images in the specified path, rather than the path provided by --root.
 
+**--included-pod-metrics**="": A list of included pod metrics. Specify the names of the metrics to be included in this list.
```
```suggestion
**--included-pod-metrics**="": A list of pod metrics names to include.
```

kwilczynski commented on docs/crio.conf.5.md:
```diff
@@ -528,7 +528,13 @@ The `crio.metrics` table containers settings pertaining to the Prometheus based
+  The number of seconds between collecting pod/container stats and pod sandbox metrics. If set to 0, the metrics/stats are collected on-demand instead.
+
+**included_pod_metrics**=[]
+  List of included pod metrics. Specify the names of the metrics to be included in this list.
```
```suggestion
  A list of pod metrics to include. Specify the names of the metrics to include in this list.
```

kwilczynski commented on internal/config/cgmgr/stats_linux.go:
```diff
@@ -37,13 +37,19 @@ type MemoryStats struct {
 	SwapLimit       uint64
+	// Amount of cached filesystem data mapped with mmap()
+	FileMapped uint64
+	// The number of memory usage hits limits. For cgroup v1 only
```
```suggestion
	// The number of memory usage hits limits. For cgroup v1 only.
```

Mo-Fatah replied:
how many times the process reached it's memory limit. in the comment above, I used the kernel [docs](https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt) definition for  `memory.failcnt`. I can change it if you like since it seems a bit vague.

kwilczynski commented on internal/config/cgmgr/stats_linux.go:
```diff
@@ -37,13 +37,19 @@ type MemoryStats struct {
 	KernelTCPUsage  uint64
 	SwapUsage       uint64
 	SwapLimit       uint64
+	// Amount of cached filesystem data mapped with mmap()
```
```suggestion
	// Amount of cached filesystem data mapped with mmap().
```

kwilczynski commented on internal/config/cgmgr/stats_linux.go:
```diff
@@ -142,6 +151,12 @@ func cgroupMemStats(memStats *libctrcgroups.MemoryStats) *MemoryStats {
 		inactiveFileName = "total_inactive_file"
 		rssBytes = memStats.Stats["total_rss"]
 		memSwap = memStats.SwapUsage.Usage
+		if memStats.UseHierarchy {
+			fileMapped = memStats.Stats["total_mapped_file"]
+		} else {
+			fileMapped = memStats.Stats["mapped_file"]
+		}
```
Perhaps?

```suggestion

		fileMapped = memStats.Stats["mapped_file"]
		if memStats.UseHierarchy {
			fileMapped = memStats.Stats["total_mapped_file"]
		}
```

kwilczynski commented on internal/lib/stats/cpu_metrics.go:
```diff
@@ -0,0 +1,113 @@
+		{
+			desc: &types.MetricDescriptor{
+				Name:      "container_cpu_user_seconds_total",
+				Help:      "Cumulative user cpu time consumed in seconds.",
```
```suggestion
				Help:      "Cumulative user CPU time consumed in seconds.",
```

kwilczynski commented on internal/lib/stats/cpu_metrics.go:
```diff
@@ -0,0 +1,113 @@
+		}, {
+			desc: &types.MetricDescriptor{
+				Name:      "container_cpu_system_seconds_total",
+				Help:      "Cumulative system cpu time consumed in seconds.",
```
```suggestion
				Help:      "Cumulative system CPU time consumed in seconds.",
```

kwilczynski commented on internal/lib/stats/cpu_metrics.go:
```diff
@@ -0,0 +1,113 @@
+		}, {
+			desc: &types.MetricDescriptor{
+				Name:      "container_cpu_usage_seconds_total",
+				Help:      "Cumulative cpu time consumed in seconds.",
```
```suggestion
				Help:      "Cumulative CPU time consumed in seconds.",
```

kwilczynski commented on internal/lib/stats/memory_metrics.go:
```diff
@@ -0,0 +1,169 @@
+		{
+			desc: &types.MetricDescriptor{
+				Name:      "container_memory_failcnt",
+				Help:      "Number of memory usage hits limits",
```
```suggestion
				Help:      "Number of memory usage hits limits.",
```

kwilczynski commented on internal/lib/stats/memory_metrics.go:
```diff
@@ -0,0 +1,169 @@
+		{
+			desc: &types.MetricDescriptor{
+				Name:      "container_memory_usage_bytes",
+				Help:      "Current memory usage in bytes, including all memory regardless of when it was accessed",
```
```suggestion
				Help:      "Current memory usage in bytes, including all memory regardless of when it was accessed.",
```

kwilczynski commented on internal/lib/stats/memory_metrics.go:
```diff
@@ -0,0 +1,169 @@
+		{
+			desc: &types.MetricDescriptor{
+				Name:      "container_memory_max_usage_bytes",
+				Help:      "Maximum memory usage recorded in bytes",
```
```suggestion
				Help:      "Maximum memory usage recorded in bytes.",
```

kwilczynski commented on internal/lib/stats/memory_metrics.go:
```diff
@@ -0,0 +1,169 @@
+		{
+			desc: &types.MetricDescriptor{
+				Name:      "container_oom_events_total",
+				Help:      "Count of out of memory events observed for the container",
```
```suggestion
				Help:      "Count of out of memory events observed for the container.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+		"cpu": {
+			{
+				Name:      "container_cpu_user_seconds_total",
+				Help:      "Cumulative user cpu time consumed in seconds.",
```
```suggestion
				Help:      "Cumulative user CPU time consumed in seconds.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+	}
+}
+
+// store metricdescriptors statically at startup, populate the list
```
The Go style comments are in the form of:

```go
// PopulateMetricDescriptors stores ...
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+				LabelKeys: baseLabelKeys,
+			}, {
+				Name:      "container_cpu_usage_seconds_total",
+				Help:      "Cumulative cpu time consumed in seconds.",
```
```suggestion
				Help:      "Cumulative CPU time consumed in seconds.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+		"cpuLoad": {
+			{
+				Name:      "container_cpu_load_average_10s",
+				Help:      "Value of container cpu load average over the last 10 seconds.",
```
```suggestion
				Help:      "Value of container CPU load average over the last 10 seconds.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+				LabelKeys: baseLabelKeys,
+			}, {
+				Name:      "container_tasks_state",
+				Help:      "Number of tasks in given state",
```
```suggestion
				Help:      "Number of tasks in given state.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+		"disk": {
+			{
+				Name:      "container_fs_inodes_free",
+				Help:      "Number of available Inodes",
```
```suggestion
				Help:      "Number of available Inodes.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+				LabelKeys: append(baseLabelKeys, "device"),
+			}, {
+				Name:      "container_fs_inodes_total",
+				Help:      "Number of Inodes",
```
Perhaps we should mention that this is a _total_ number of inodes?

Mo-Fatah replied:
changed it to total. I guess no problem if it slightly differs from cAdvisor (@haircommander ?)

haircommander replied:
I vote consistency with cadvisor https://github.com/google/cadvisor/blob/54dff2b8ccb147747d814b0ff3b4a3256dc569c0/metrics/prometheus.go#L517

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+		"diskio": {
+			{
+				Name:      "container_fs_reads_bytes_total",
+				Help:      "Cumulative count of bytes read",
```
```suggestion
				Help:      "Cumulative count of bytes read.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_fs_reads_total",
+				Help:      "Cumulative count of reads completed",
```
```suggestion
				Help:      "Cumulative count of reads completed.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_fs_sector_reads_total",
+				Help:      "Cumulative count of sector reads completed",
```
```suggestion
				Help:      "Cumulative count of sector reads completed.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_fs_reads_merged_total",
+				Help:      "Cumulative count of reads merged",
```
```suggestion
				Help:      "Cumulative count of reads merged.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_fs_read_seconds_total",
+				Help:      "Cumulative count of seconds spent reading",
```
```suggestion
				Help:      "Cumulative count of seconds spent reading.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_fs_writes_bytes_total",
+				Help:      "Cumulative count of bytes written",
```
```suggestion
				Help:      "Cumulative count of bytes written.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_fs_writes_total",
+				Help:      "Cumulative count of writes completed",
```
```suggestion
				Help:      "Cumulative count of writes completed.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_fs_sector_writes_total",
+				Help:      "Cumulative count of sector writes completed",
```
```suggestion
				Help:      "Cumulative count of sector writes completed.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_fs_writes_merged_total",
+				Help:      "Cumulative count of writes merged",
```
```suggestion
				Help:      "Cumulative count of writes merged.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_fs_write_seconds_total",
+				Help:      "Cumulative count of seconds spent writing",
```
```suggestion
				Help:      "Cumulative count of seconds spent writing.",
```

haircommander replied:
naturally there's no consistency in cadvisor on whether to end with a period: https://github.com/google/cadvisor/blob/master/metrics/prometheus.go#L674 I think we should keep these consistent with the cadvisor equivalent for now, to allow easier comparison

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_fs_io_current",
+				Help:      "Number of I/Os currently in progress",
```
```suggestion
				Help:      "Number of I/Os currently in progress.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_fs_io_time_seconds_total",
+				Help:      "Cumulative count of seconds spent doing I/Os",
```
```suggestion
				Help:      "Cumulative count of seconds spent doing I/Os.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_fs_io_time_weighted_seconds_total",
+				Help:      "Cumulative weighted I/O time in seconds",
```
```suggestion
				Help:      "Cumulative weighted I/O time in seconds.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_blkio_device_usage_total",
+				Help:      "Blkio Device bytes usage",
```
```suggestion
				Help:      "Blkio device bytes usage.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_memory_failcnt",
+				Help:      "Number of memory usage hits limits",
```
```suggestion
				Help:      "Number of memory usage hits limits.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_memory_usage_bytes",
+				Help:      "Current memory usage in bytes, including all memory regardless of when it was accessed",
```
```suggestion
				Help:      "Current memory usage in bytes, including all memory regardless of when it was accessed.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_memory_max_usage_bytes",
+				Help:      "Maximum memory usage recorded in bytes",
```
```suggestion
				Help:      "Maximum memory usage recorded in bytes.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+		"misc": {
+			{
+				Name:      "container_scrape_error",
+				Help:      "1 if there was an error while getting container metrics, 0 otherwise",
```
```suggestion
				Help:      "Set to 1 if there was an error while getting container metrics, 0 otherwise.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+				LabelKeys: baseLabelKeys,
+			}, {
+				Name:      "container_last_seen",
+				Help:      "Last time a container was seen by the exporter",
```
```suggestion
				Help:      "Last time a container was seen by the exporter.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+				LabelKeys: append(baseLabelKeys, "interface"),
+			}, {
+				Name:      "container_network_receive_packets_total",
+				Help:      "Cumulative count of packets received",
```
```suggestion
				Help:      "Cumulative count of packets received.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+				LabelKeys: append(baseLabelKeys, "interface"),
+			}, {
+				Name:      "container_network_receive_packets_dropped_total",
+				Help:      "Cumulative count of packets dropped while receiving",
```
```suggestion
				Help:      "Cumulative count of packets dropped while receiving.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+				LabelKeys: append(baseLabelKeys, "interface"),
+			}, {
+				Name:      "container_network_receive_errors_total",
+				Help:      "Cumulative count of errors encountered while receiving",
```
```suggestion
				Help:      "Cumulative count of errors encountered while receiving.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+				LabelKeys: append(baseLabelKeys, "interface"),
+			}, {
+				Name:      "container_network_transmit_bytes_total",
+				Help:      "Cumulative count of bytes transmitted",
```
```suggestion
				Help:      "Cumulative count of bytes transmitted.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+				LabelKeys: append(baseLabelKeys, "interface"),
+			}, {
+				Name:      "container_network_transmit_packets_total",
+				Help:      "Cumulative count of packets transmitted",
```
```suggestion
				Help:      "Cumulative count of packets transmitted.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+				LabelKeys: append(baseLabelKeys, "interface"),
+			}, {
+				Name:      "container_network_transmit_packets_dropped_total",
+				Help:      "Cumulative count of packets dropped while transmitting",
```
```suggestion
				Help:      "Cumulative count of packets dropped while transmitting.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+				LabelKeys: append(baseLabelKeys, "interface"),
+			}, {
+				Name:      "container_network_transmit_errors_total",
+				Help:      "Cumulative count of errors encountered while transmitting",
```
```suggestion
				Help:      "Cumulative count of errors encountered while transmitting.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+		"oom": {
+			{
+				Name:      "container_oom_events_total",
+				Help:      "Count of out of memory events observed for the container",
```
```suggestion
				Help:      "Count of out of memory events observed for the container.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_threads_max",
+				Help:      "Maximum number of threads allowed inside the container, infinity if value is zero",
```
```suggestion
				Help:      "Maximum number of threads allowed inside the container, infinity if value is zero.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_threads",
+				Help:      "Number of threads running inside the container",
```
```suggestion
				Help:      "Number of threads running inside the container.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			},
+			{
+				Name:      "container_ulimits_soft",
+				Help:      "Soft ulimit values for the container root process. Unlimited if -1, except priority and nice",
```
```suggestion
				Help:      "Soft ulimit values for the container root process. Unlimited if '-1', except priority and nice.",
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+	for _, m := range metrics {
+		metricValues := m.valueFunc()
+		if len(metricValues) == 0 {
+			// No metrics to process for this ContainerMetrics, move to the next one
```
No need to explain obvious code.

```suggestion
```

kwilczynski commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,368 @@
+			// No metrics to process for this ContainerMetrics, move to the next one
+			continue
+		}
+		for _, v := range metricValues {
```
`range` does not iterate over an empty slice. Thus, perhaps?

```suggestion
		for _, v := range m.valueFunc() {
```

Mo-Fatah replied:
yes. much cleaner

kwilczynski commented on internal/lib/stats/network_metrics.go:
```diff
@@ -0,0 +1,134 @@
+		return nil
+	}
+	if len(links) == 0 {
+		logrus.Infof("Network links are not available.")
```
Perhaps this warrants a warning?

```suggestion
		logrus.Warn("Network links are not available.")
```

kwilczynski commented on internal/lib/stats/network_metrics.go:
```diff
@@ -0,0 +1,134 @@
+		return nil
+	}
+	for i := range links {
+		attrs := links[i].Attrs()
+		if attrs != nil {
```
Consistency in style.

```suggestion
		if attrs := links[i].Attrs(); attrs != nil {
```

kwilczynski commented on internal/lib/stats/network_metrics.go:
```diff
@@ -0,0 +1,134 @@
+		{
+			desc: &types.MetricDescriptor{
+				Name:      "container_network_receive_bytes_total",
+				Help:      "Cumulative count of bytes received",
```
```suggestion
				Help:      "Cumulative count of bytes received.",
```

kwilczynski commented on internal/lib/stats/network_metrics.go:
```diff
@@ -0,0 +1,134 @@
+		}, {
+			desc: &types.MetricDescriptor{
+				Name:      "container_network_receive_packets_total",
+				Help:      "Cumulative count of packets received",
```
```suggestion
				Help:      "Cumulative count of packets received.",
```

kwilczynski commented on internal/lib/stats/network_metrics.go:
```diff
@@ -0,0 +1,134 @@
+		}, {
+			desc: &types.MetricDescriptor{
+				Name:      "container_network_receive_packets_dropped_total",
+				Help:      "Cumulative count of packets dropped while receiving",
```
```suggestion
				Help:      "Cumulative count of packets dropped while receiving.",
```

kwilczynski commented on internal/lib/stats/network_metrics.go:
```diff
@@ -0,0 +1,134 @@
+		}, {
+			desc: &types.MetricDescriptor{
+				Name:      "container_network_receive_errors_total",
+				Help:      "Cumulative count of errors encountered while receiving",
```
```suggestion
				Help:      "Cumulative count of errors encountered while receiving.",
```

kwilczynski commented on internal/lib/stats/network_metrics.go:
```diff
@@ -0,0 +1,134 @@
+		}, {
+			desc: &types.MetricDescriptor{
+				Name:      "container_network_transmit_bytes_total",
+				Help:      "Cumulative count of bytes transmitted",
```
```suggestion
				Help:      "Cumulative count of bytes transmitted.",
```

kwilczynski commented on internal/lib/stats/network_metrics.go:
```diff
@@ -0,0 +1,134 @@
+		}, {
+			desc: &types.MetricDescriptor{
+				Name:      "container_network_transmit_packets_total",
+				Help:      "Cumulative count of packets transmitted",
```
```suggestion
				Help:      "Cumulative count of packets transmitted.",
```

kwilczynski commented on internal/lib/stats/network_metrics.go:
```diff
@@ -0,0 +1,134 @@
+		}, {
+			desc: &types.MetricDescriptor{
+				Name:      "container_network_transmit_packets_dropped_total",
+				Help:      "Cumulative count of packets dropped while transmitting",
```
```suggestion
				Help:      "Cumulative count of packets dropped while transmitting.",
```

kwilczynski commented on internal/lib/stats/stats_server.go:
```diff
@@ -228,3 +230,31 @@ func (ss *StatsServer) Shutdown() {
+	return ss.metricsForPodSandbox(sb)
+}
+
+// MetricsForPodSandboxList returns the metrics for the given list of sandboxes
```
```suggestion
// MetricsForPodSandboxList returns the metrics for the given list of sandboxes.
```

kwilczynski commented on internal/lib/stats/stats_server.go:
```diff
@@ -228,3 +230,31 @@ func (ss *StatsServer) Shutdown() {
+}
+
+// RemoveMetricsForPodSandbox removes the saved entry for the specified sandbox
+// to prevent the map from always growing.
```
Mind you, given the internal implementation of Go's map, this does not stop it from growing. You don't have much control over the map's growth and when garbage will be collected unless you are very explicit about memory management.

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -21,6 +22,14 @@ func (ss *StatsServer) updateSandbox(sb *sandbox.Sandbox) *types.PodSandboxStats
 		return nil
 	}
+
+	// Sandbox metrics are to fulfill the CRI metrics endpoint
```
```suggestion
	// Sandbox metrics are to fulfill the CRI metrics endpoint.
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -21,6 +22,14 @@ func (ss *StatsServer) updateSandbox(sb *sandbox.Sandbox) *types.PodSandboxStats
+		sandboxMetrics = NewSandboxMetrics(sb)
+	}
+
+	// Sandbox stats are to fulfill the Kubelet's /stats/summary endpoint
```
```suggestion
	// Sandbox stats are to fulfill the Kubelet's /stats/summary endpoint.
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -31,6 +40,12 @@ func (ss *StatsServer) updateSandbox(sb *sandbox.Sandbox) *types.PodSandboxStats
 		Linux: &types.LinuxPodSandboxStats{},
 	}
 
+	// network metrics are collected at pod level only
```
```suggestion
	// Network metrics are collected at pod level only.
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -42,8 +57,13 @@ func (ss *StatsServer) updateSandbox(sb *sandbox.Sandbox) *types.PodSandboxStats
-	for _, c := range sb.Containers().List() {
+
+	containersList := sb.Containers().List()
+
```
```suggestion
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -52,25 +72,36 @@ func (ss *StatsServer) updateSandbox(sb *sandbox.Sandbox) *types.PodSandboxStats
 		}
 		containerStats = append(containerStats, cStats)
+
+		// convert cgroups stats to CRI metrics
```
```suggestion
		// Convert cgroups stats to CRI metrics.
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -52,25 +72,36 @@ func (ss *StatsServer) updateSandbox(sb *sandbox.Sandbox) *types.PodSandboxStats
 			logrus.Errorf("Error getting container stats %s: %v", c.ID(), err)
 			continue
 		}
+		// convert cgroups stats to CRI stats
```
```suggestion
		// Convert cgroups stats to CRI stats.
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
+	if ok {
+		return sboxMetrics
+	}
+	// Cache miss, try again
```
```suggestion
	// Cache miss, try again.
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
+	if ss.collectionPeriod == 0 {
+		return ss.updatePodSandboxMetrics(sb)
+	}
+	sboxMetrics, ok := ss.sboxMetrics[sb.ID()]
+	if ok {
```
```suggestion
	if sboxMetrics, ok := ss.sboxMetrics[sb.ID()]; ok {
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
+	if !exists {
+		sm = NewSandboxMetrics(sb)
+	}
+	// network metrics are collected at pod level
```
```suggestion
	//Network metrics are collected at the pod level.
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
+	containersList := sb.Containers().List()
+	containerMetrics := make([]*types.ContainerMetrics, 0, len(containersList))
+	for _, c := range containersList {
+		// Skip if the container is stopped
```
```suggestion
		// Skip if the container is stopped.
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
+		containerMetrics = append(containerMetrics, cMetrics)
+	}
+	sm.metric.ContainerMetrics = containerMetrics
+	ss.sboxMetrics[sb.ID()] = sm // Update the entry in the map
```
No need to explain obvious code.

```suggestion
	ss.sboxMetrics[sb.ID()] = sm
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
+	return sm
+}
+
+// GenerateSandboxContainerMetrics generates a list of metrics for the specified sandbox
+// containers by collecting metrics from the cgroup based on the includedMetrics
+// except for network metrics which is collected at pod level.
```
```suggestion
// GenerateSandboxContainerMetrics generates a list of metrics for the specified sandbox
// containers by collecting metrics from the cgroup based on the includedMetrics
// except for network metrics, which are collected at the pod level.
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
+			}
+			oomCount, err := cm.OOMKillCount()
+			if err != nil {
+				logrus.Errorf("Unable to fetch oom kill count %s: %v", sb.ID(), err)
```
```suggestion
				logrus.Errorf("Unable to fetch OOM kill count %s: %v", sb.ID(), err)
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
+		case "oom":
+			cm, err := ss.Config().CgroupManager().ContainerCgroupManager(sb.CgroupParent(), c.ID())
+			if err != nil {
+				logrus.Errorf("Unable to fetch cgroup manager %s: %v", sb.ID(), err)
```
Albeit, I am not sure if saying "container" rather than "sandbox" would be more correct here?

Mo-Fatah replied:
yeah, container is correct here.

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
+			oomMetrics := GenerateSandboxOOMMetrics(sb, c, oomCount)
+			metrics = append(metrics, oomMetrics...)
+		case "network":
+			continue // network metrics are collected at pod level only
```
```suggestion
			continue // Network metrics are collected at the pod level.
```

kwilczynski commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
+		case "network":
+			continue // network metrics are collected at pod level only
+		default:
+			logrus.Warnf("Unknown or misspelled metric: %s", m)
```
Misspelled metrics are unknown metrics.

```suggestion
			logrus.Warnf("Unknown metric: %s", m)
```

kwilczynski commented on internal/lib/stats/stats_server_unsupported.go:
```diff
@@ -19,6 +19,13 @@ func (ss *StatsServer) updateSandbox(sb *sandbox.Sandbox) *types.PodSandboxStats
+
+// metricsForPodSandbox is an internal, non-locking version of MetricsForPodSandbox
+// that returns (and occasionally gathers) the metrics for the given sandbox.
+// Note: caller must hold the lock on the StatsServer
```
```suggestion
// Note: the caller must hold the lock on the StatsServer.
```

kwilczynski commented on server/metric_descriptors_list.go:
```diff
@@ -2,12 +2,23 @@ package server
+	includedKeys := []string{"cpu", "memory", "network"} // You can modify this as needed
+	descriptorsMap := s.StatsServer.PopulateMetricDescriptors(includedKeys)
+
+	// Flatten the map of descriptors to a slice
```
```suggestion
	// Flatten the map of descriptors to a slice.
```

kwilczynski commented on server/metric_descriptors_list.go:
```diff
@@ -2,12 +2,23 @@ package server
+		flattenedDescriptors = append(flattenedDescriptors, descriptors...)
+	}
+
+	// Create the response with the flattened list of descriptors
```
```suggestion
```

kwilczynski commented on server/sandbox_metrics_list.go:
```diff
@@ -2,12 +2,32 @@ package server
+		if current := metrics.GetMetric(); current != nil {
+			responseMetricsList = append(responseMetricsList, current)
+		} else {
+			// Iterate over container metrics within each PodSandboxMetrics
```
```suggestion
			// Iterate over container metrics within each PodSandboxMetrics.
```

kwilczynski commented on test/cri-metrics.bats:
```diff
@@ -0,0 +1,145 @@
+    sleep 1
+
+    # assert container_memory_usage_bytes == cgroup memory.usage_in_bytes(cgroup v1) or memory.current(cgroup v2)
+    if is_cgroup_v2; then
```
Normally, you would have two tests, one for cgroup v1 and the other for v2, and you would skip its execution based on whether you are dealing with v1 or v2 capable platform.

This would remove clutter from the test, even though there would be some duplication.

kwilczynski commented on completions/fish/crio.fish:
```diff
@@ -24,6 +24,7 @@ complete -c crio -n '__fish_crio_no_subcommand' -l clean-shutdown-file -r -d 'Lo
 complete -c crio -n '__fish_crio_no_subcommand' -l cni-config-dir -r -d 'CNI configuration files directory.'
 complete -c crio -n '__fish_crio_no_subcommand' -f -l cni-default-network -r -d 'Name of the default CNI network to select. If not set or "", then CRI-O will pick-up the first one found in --cni-config-dir.'
 complete -c crio -n '__fish_crio_no_subcommand' -f -l cni-plugin-dir -r -d 'CNI plugin binaries directory.'
+complete -c crio -n '__fish_crio_no_subcommand' -f -l collection-period -r -d 'The number of seconds between collecting pod/container stats and pod sandbox metrics. If set to 0, the metrics/stats are collected on-demand instead.'
```
Are these metrics or stats collected for pod or a container? What is the difference between a stat and a metric?

kwilczynski commented on completions/fish/crio.fish:
```diff
@@ -89,6 +90,7 @@ complete -c crio -n '__fish_crio_no_subcommand' -f -l image-volumes -r -d 'Image
        mounted into the container for the volumes.
 	3. ignore: All volumes are just ignored and no action is taken.'
 complete -c crio -n '__fish_crio_no_subcommand' -l imagestore -r -d 'Store newly pulled images in the specified path, rather than the path provided by --root.'
+complete -c crio -n '__fish_crio_no_subcommand' -f -l included-pod-metrics -r -d 'A list of included pod metrics. Specify the names of the metrics to be included in this list.'
```
See question above.

kwilczynski commented on completions/fish/crio.fish:
```diff
@@ -151,7 +153,7 @@ complete -c crio -n '__fish_crio_no_subcommand' -f -l separate-pull-cgroup -r -d
 complete -c crio -n '__fish_crio_no_subcommand' -l signature-policy -r -d 'Path to signature policy JSON file.'
 complete -c crio -n '__fish_crio_no_subcommand' -l signature-policy-dir -r -d 'Path to the root directory for namespaced signature policies. Must be an absolute path.'
-complete -c crio -n '__fish_crio_no_subcommand' -f -l stats-collection-period -r -d 'The number of seconds between collecting pod and container stats. If set to 0, the stats are collected on-demand instead.'
+complete -c crio -n '__fish_crio_no_subcommand' -f -l stats-collection-period -r -d 'The number of seconds between collecting pod and container stats. If set to 0, the stats are collected on-demand instead. This option is deprecated and will be removed in the future.'
```
```suggestion
complete -c crio -n '__fish_crio_no_subcommand' -f -l stats-collection-period -r -d 'The number of seconds between collecting pod and container stats. If set to 0, the stats are collected on-demand instead. DEPRECATED: This option will be removed in the future.'
```

kwilczynski commented on completions/bash/crio:
```diff
@@ -29,6 +29,7 @@ h
 --cni-config-dir
 --cni-default-network
 --cni-plugin-dir
+--collection-period
```
Perhaps? Not entirely sure, though.

```suggestion
--metrics-collection-period
```

Mo-Fatah replied:
I guess `collection-period` is much more suitable since it refers to both stats and metrics collection. see https://github.com/cri-o/cri-o/pull/7796#discussion_r1567830047

kwilczynski commented on completions/zsh/_crio:
```diff
@@ -36,6 +36,7 @@ it later with **--config**. Global options will modify the output.'
         '--cni-config-dir'
         '--cni-default-network'
         '--cni-plugin-dir'
+        '--collection-period'
```
See question above.

kwilczynski commented on completions/zsh/_crio:
```diff
@@ -69,6 +70,7 @@ it later with **--config**. Global options will modify the output.'
         '--hostnetwork-disable-selinux'
         '--image-volumes'
         '--imagestore'
+        '--included-pod-metrics'
```
See question above.

kwilczynski commented on docs/crio.8.md:
```diff
@@ -26,6 +26,7 @@ crio
 [--cni-config-dir]=[value]
 [--cni-default-network]=[value]
 [--cni-plugin-dir]=[value]
+[--collection-period]=[value]
```
Ditto.

kwilczynski commented on docs/crio.8.md:
```diff
@@ -60,6 +61,7 @@ crio
 [--hostnetwork-disable-selinux]
 [--image-volumes]=[value]
 [--imagestore]=[value]
+[--included-pod-metrics]=[value]
```
Ditto.

kwilczynski commented on internal/lib/stats/cpu_metrics.go:
```diff
@@ -0,0 +1,113 @@
+	cpu, ok := stats.(*cgmgr.CPUStats)
+	if !ok {
+		logrus.Errorf("Failed to assert stats as *cgmgr.CpuStats")
+		return nil
```
If this ever fails, you will append `nil` into your metrics slice, and then it will be passed all over the place, before something tries to iterate over it and use the value to call some callable, and then it will explode, I believe.

Failing a type assertion might carry more weight than just a log line, I think.

Mo-Fatah replied:
I added an if condition to guard against appending `nil` into metrics slice. I guess this should work?

kwilczynski commented on internal/lib/stats/memory_metrics.go:
```diff
@@ -0,0 +1,169 @@
+func GenerateSandboxMemoryMetrics(sb *sandbox.Sandbox, c *oci.Container, stats interface{}, sm *SandboxMetrics) []*types.Metric {
+	mem, ok := stats.(*cgmgr.MemoryStats)
+	if !ok {
+		logrus.Errorf("Failed to assert stats as *cgmgr.MemoryStats")
```
See my comment elsewhere about this.

kwilczynski commented:
@Mo-Fatah, looks very good!

I am unsure if you want to have periods at the end of the help messages for each metric, but consistency would be nice, whichever way you prefer. I went with the added period, but feel free to revise as needed.

Some of the actual help messages are a bit... terse, and could be improved. But I appreciate that some are awkward to explain.

kwilczynski commented:
/unhold

Sorry for being late with my review!

haircommander commented on completions/bash/crio:
```diff
@@ -62,6 +63,7 @@ h
 --hostnetwork-disable-selinux
 --image-volumes
 --imagestore
+--included-pod-metrics
```
I prefer included here actually

haircommander replied:
Include sounds like a boolean

Mo-Fatah commented:
@haircommander @kwilczynski please have another look when you have time, I addressed most of your comments and responded to the unresolved ones. sorry for being late on that

haircommander commented:
/approve

I think there's still more work to be done here, but you've done some great work so far. Thanks for your patience and time here @Mo-Fatah , really impressive work

@cri-o/cri-o-maintainers PTAL

kwilczynski commented:
> I think there's still more work to be done here, but you've done some great work so far.
> [...]

I second that! Thank you a lot @Mo-Fatah!

kwilczynski commented:
/approve
/lgtm

kwilczynski commented:
/hold

For @saschagrunert and @mrunalp.

kwilczynski commented:
/assign Mo-Fatah 
/assign haircommander 
/assign kwilczynski

saschagrunert commented on internal/lib/stats/cpu_metrics.go:
```diff
@@ -0,0 +1,113 @@
+	types "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+func GenerateSandboxCPUMetrics(sb *sandbox.Sandbox, c *oci.Container, stats interface{}) []*types.Metric {
```
No need to have `stats` being an `interface{}` here, please use a concrete type.

saschagrunert commented on internal/lib/stats/memory_metrics.go:
```diff
@@ -0,0 +1,169 @@
+	types "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+func GenerateSandboxMemoryMetrics(sb *sandbox.Sandbox, c *oci.Container, stats interface{}, sm *SandboxMetrics) []*types.Metric {
```
Please make private if not used in another package. `stats` can be `*cgmgr.MemoryStats` instead of `interface{}`.

saschagrunert commented on internal/lib/stats/cpu_metrics.go:
```diff
@@ -0,0 +1,113 @@
+	cpuMetrics := []*containerMetric{
+		{
+			desc: &types.MetricDescriptor{
+				Name:      "container_cpu_user_seconds_total",
+				Help:      "Cumulative user CPU time consumed in seconds.",
```
The names and corresponding help messages should be const values and reused in `PopulateMetricDescriptors`.

saschagrunert commented on internal/lib/stats/metrics.go:
```diff
@@ -0,0 +1,363 @@
+	return []string{sb.ID(), "POD", ""}
+}
+
+// ComputeSandboxMetrics computes the metrics for both pod and container sandbox.
+func ComputeSandboxMetrics(sb *sandbox.Sandbox, c *oci.Container, metrics []*containerMetric, metricName string) []*types.Metric {
```
Let's make this private.

saschagrunert commented on internal/lib/stats/stats_server_linux.go:
```diff
@@ -119,6 +149,101 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
+// containers by collecting metrics from the cgroup based on the included pod metrics,
+// except for network metrics, which are collected at the pod level.
+func (ss *StatsServer) GenerateSandboxContainerMetrics(sb *sandbox.Sandbox, c *oci.Container, sm *SandboxMetrics) *types.ContainerMetrics {
+	cgstats, err := ss.Runtime().ContainerStats(context.TODO(), c, sb.CgroupParent())
+	if err != nil || cgstats == nil {
+		logrus.Errorf("Error getting sandbox stats %s: %v", sb.ID(), err)
+		return nil
```
We should pass around the context here and use our internal `log` package everywhere.

sohankunkerkar commented:
Integration test is failing: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/cri-o_cri-o/7796/pull-ci-cri-o-cri-o-main-ci-cgroupv2-integration/1783901353014202368/artifacts/cgroupv2-integration/cri-o-gather/artifacts/testout.txt

Mo-Fatah commented:
/retest

Mo-Fatah commented:
/retest

Mo-Fatah commented:
/retest

kwilczynski commented:
/retest

Mo-Fatah commented:
/retest

kwilczynski commented on internal/lib/stats/cpu_metrics.go:
```diff
@@ -0,0 +1,105 @@
+				LabelKeys: append(baseLabelKeys, "cpu"),
+			},
+			valueFunc: func() metricValues {
+				if len(cpu.PerCPUUsage) == 0 {
+					if cpu.TotalUsageNano > 0 {
+						return metricValues{{
+							value:      cpu.TotalUsageNano / uint64(time.Second),
+							labels:     []string{"total"},
+							metricType: types.MetricType_COUNTER,
+						}}
+					}
+				}
```
Collapse to create an association for both statements to reduce indentation, perhaps?

```suggestion
				if len(cpu.PerCPUUsage) == 0 && cpu.TotalUsageNano > 0 {
					return metricValues{{
						value:      cpu.TotalUsageNano / uint64(time.Second),
						labels:     []string{"total"},
						metricType: types.MetricType_COUNTER,
					}}
				}
```

Mo-Fatah replied:
done

kwilczynski commented on test/cri-metrics.bats:
```diff
@@ -0,0 +1,162 @@
+load helpers
+
+function setup() {
+	if [[ "$CONTAINER_DEFAULT_RUNTIME" == "containerd-shim-kata-v2" ]]; then
+		# the tests read the cgroup files directly, which is not possible with containerd-shim-kata-v2; skip for now.
+		skip
+	fi
```
We also skip tests for Kata via:

- [cri-o/cri-o/contrib/test/ci/vars.yml](https://github.com/cri-o/cri-o/blob/main/contrib/test/ci/vars.yml)

Would that work for you?

Mo-Fatah replied:
yeah, I guess this is what I was trying to do. Thank you!

haircommander commented on pkg/annotations/annotations.go:
```diff
@@ -109,7 +109,7 @@ var AllAllowedAnnotations = []string{
 	// Keep in sync with
 	// https://github.com/opencontainers/runc/blob/3db0871f1cf25c7025861ba0d51d25794cb21623/features.go#L67
-	// Once runc 1.2 is released, we can use the `runc features` command to get this programatically,
+	// Once runc 1.2 is released, we can use the `runc features` command to get this programmatically,
```
would you mind dropping this and rebaseing on https://github.com/cri-o/cri-o/pull/8088

Mo-Fatah replied:
:heavy_check_mark:

haircommander commented on test/workloads.bats:
```diff
@@ -402,7 +402,7 @@ function check_conmon_fields() {
 
 	ctr_id=$(crictl run "$ctrconfig" "$sboxconfig")
-	[[ $(systemctl show --property CollectMode crio-$ctr_id.scope) != "CollectMode=inactive-or-failed" ]]
+	[[ $(systemctl show --property CollectMode crio-"$ctr_id".scope) != "CollectMode=inactive-or-failed" ]]
```
and drop this one as well please

haircommander commented:
/override ci/prow/ci-e2e-evented-pleg
/lgtm

Thanks @Mo-Fatah ! Great work here

saschagrunert commented:
/retest

Mo-Fatah commented:
/retest

haircommander commented:
/lgtm
/approve

:tada:

haircommander commented:
/retest

haircommander commented:
/retest

kwilczynski commented:
/test e2e-gcp-ovn

Mo-Fatah commented:
/retest

haircommander commented:
/override ci/prow/e2e-gcp-ovn
